core/include/kernel/pm.h

```diff
@@ -0,0 +1,137 @@
+ * PM_OP_RESUME: platform is resuming from low power state
+ */
+enum pm_op {
+	PM_OP_SUSPEND = 0,
+	PM_OP_RESUME = 1,
+};
+
+/*
+ * Registered callbacks are called the ordering directives
+ * specified by the PM_CB_LEVEL_xxx value.
+ */
+enum pm_callback_order {
+	PM_CB_ORDER_DRIVER = 0,
+	PM_CB_ORDER_CORE_SERVICE,
+	PM_CB_ORDER_MAX
+};
+
+#define PM_CALLBACK_HANDLE_INITIALIZER(_callback, _handle, _order)	\
+		(struct pm_callback_handle){				\
+			.callback = (_callback),			\
+			.handle = (_handle),				\
+			.order = (_order),				\
+		}
+
+#define PM_CALLBACK_GET_HANDLE(pm_handle)	((pm_handle)->handle)
+
+/*
+ * Drivers and services can register a callback function for the platform
+ * suspend and resume sequences. A private address handle can be registered
+ * with the callback and retrieved from the callback. Callback can be
+ * registered with a specific call order as defined per PM_CB_ORDER_*.
+ *
+ * Callback shall return an error if failing to complete target transition.
+ * This information may be used by the platform to resume a platform on
+ * non-fatal failure to suspend.
+ *
+ * Callback function API:
+ *
+ * TEE_Result (*callback)(enum pm_op op,
+ *			  unsigned int pm_hint,
+ *			  const struct pm_callback_handle *pm_handle);
+ *
+ * @op - Target operation: either PM_SUSPEND or PM_RESUME
+ * @pm_hint - Hints on power state platform suspends to /resumes from.
+ *		PM_STATE_HINT_* defines the supported values.
+ * @pm_handle - Reference to the struct pm_callback_handle related to to
+ *		registered callback. Callback can retrieve the registered
+ *		private handle with PM_CALLBACK_GET_HANDLE().
+ *
+ * Return a TEE_Result compliant return code
+ */
+/*
+ * struct pm_callback_handle store the callback registration directives.
+ *
+ * @callback - Registered callback function
+ * @handle - Registered private handler for the callback
+ * @order - Registered callback call order priority (PM_CB_ORDER_*)
+ */
+struct pm_callback_handle {
+	/* Set by the caller when registering a callback */
+	TEE_Result (*callback)(enum pm_op op, uint32_t pm_hint,
+			       const struct pm_callback_handle *pm_handle);
+	void *handle;
+	uint8_t order;
+	/* Set by the system according to execution context */
+	uint8_t flags;
+};
+
+/*
+ * Register a callback for suspend/resume sequence
+ * Refer to struct pm_callback_handle for description of the callbacks
+ * API and the registration directives.
+ *
+ * @pm_hdl: Reference callback registration directive
+ *
+ * Driver should tag its callback resource as unpaged (i.e KEEP_PAGER())
+ * since the callback is called from an unpaged execution context.
+ */
+void register_pm_cb(struct pm_callback_handle *pm_hdl);
+
+/*
+ * Register a driver callback for generic suspend/resume.
+ * Refer to struct pm_callback_handle for description of the callbacks
+ * API.
+ *
+ * @callback: Registered callback function
+ * @handle: Registered private handle argument for the callback
+ */
+void register_pm_driver_cb(
```

Comment:
Ok, I will make it as an inline function. 
As a driver writer, I would prefer to implementing something like 
```c
	(...)
	register_pm_driver_cb(gpio_secure_suspend_resume, dev);
	(...)
```
rather than
```c
	(...)
	register_pm_cb(&PM_CALLBACK_HANDLE_INITIALIZER(
				gpio_secure_suspend_resume, dev,
				PM_CB_ORDER_DRIVER));
	(...)
```

Reply:
Fair enough.